### DISCUSSION TITLE
align parser hooks with webpack: multi hooks in one visitor instead of multi visitor

### DISCUSSION THREAD

**Discussion by ahabhgk** [web-infra-dev/rspack#3619] [Core]
Some benefits
- one visit, better performance
- decouple some visitor logic, for example [`UrlScanner` and `WorkerScanner`](https://github.com/web-infra-dev/rspack/pull/3617/files#diff-4a3284780447bbeee1c44bcbee168bd68eec2926a6de6697ca74f3c4bef2a6c0R27-R30)

Related functionality (might not dependent on it but currently rspack don't have)
- [tagVariable](https://github.com/webpack/webpack/blob/1e18b1d1c74e26a84e9adee519c045e2d8ed40a4/lib/dependencies/WorkerPlugin.js#L450-L469)
- [compile time evaluate](https://github.com/webpack/webpack/blob/1e18b1d1c74e26a84e9adee519c045e2d8ed40a4/lib/dependencies/WorkerPlugin.js#L142)

TODO...

**Comment by hyf0** [comment]
>
> * one visit, better performance
> * decouple some visitor logic, for example [`UrlScanner` and `WorkerScanner`](https://github.com/web-infra-dev/rspack/pull/3617/files#diff-4a3284780447bbeee1c44bcbee168bd68eec2926a6de6697ca74f3c4bef2a6c0R27-R30)
> 

I'm going to implement the `parser hooks` on the rust side. Many scanner logic depends on this. if we don't do it now, it would be much more code to be refactored in the future.
